src/client/mapblock_mesh.cpp

```diff
@@ -1582,3 +1582,31 @@ video::SColor encode_light(u16 light, u8 emissive_light)
 	float b = (day + night) / 2;
 	return video::SColor(r, b, b, b);
 }
+
+u8 get_solid_sides(MeshMakeData *data)
+{
+	v3s16 blockpos_nodes = data->m_blockpos * MAP_BLOCKSIZE;
+	const NodeDefManager *ndef = data->m_client->ndef();
+
+	u8 result = 0x3F; // all sides solid;
+
+	for (s16 i = 0; i < MAP_BLOCKSIZE && result != 0; i++)
+	for (s16 j = 0; j < MAP_BLOCKSIZE && result != 0; j++) {
+		v3s16 positions[6] = {
+			v3s16(0, i, j),
+			v3s16(MAP_BLOCKSIZE - 1, i, j),
+			v3s16(i, 0, j),
+			v3s16(i, MAP_BLOCKSIZE - 1, j),
+			v3s16(i, j, 0),
+			v3s16(i, j, MAP_BLOCKSIZE - 1)
+		};
+
+		for (u8 k = 0; k < 6; k++) {
+			const MapNode &top = data->m_vmanip.getNodeRefUnsafe(blockpos_nodes + positions[k]);
+			if (ndef->get(top).solidness != 2)
+				result &= ~(1 << k);
+		}
+	}
+
+	return result;
```

Comment:
This code checks the entire cube, not the sides.

Reply:
The x-sidelength of the axis-aligned box with corners `v3s16(0, 0, 0), v3s16(1, B, B)` is `1-0 = 1`. This is the x-side.

(There was a mistake in my suggestion (volume=0), fixed that.)

Reply:
I am not sure the suggestion is more optimal or is easier to read. I blended the loops exactly to avoid the overhead of repeating for-for loops 6 times, and the inner loop of 6 iterations is easy for compiler to unroll.

Reflecting on what people mostly play, the early breaks may not be a significant gain, because 40-99% of the blocks around the player are solid or have mostly full solid sides, unless the player is flying high in the skies.

I can change the (result != 0) condition into something more readable if that's a problem.

Reply:
Fine then. This code is also not that performance critical anyway, I guess.